design/automatic-network-configuration.md

```diff
@@ -0,0 +1,641 @@
+
+## Non-Goals
+
+- Implement a network controller which is aimed at being an integration with
+  an existing controller only.
+- Implement a solution for a specific underlying infrastructure.
+- Implement a solution for link aggregation.
+- It is not a goal (of this spec) to configure the software running on the Host to
+  connect to the right networks on the right ports.
+
+## Proposal
+
+This document proposes to add a new mechanic to automatically perform physical
+network device configuration when consuming a bmh.
+There will be a new operator and a series of CRDs staying in a new separate
+repository, and relative changes to CAPM3 and BMO.
+
+### User Stories
+
+#### Story 1
+
+As the administrator of Metal³, I hope that the machine can be automatically
+connected to the provisioning network before provisioning.
+
+#### Story 2
+
+As a administrator of Metal³, I hope that the machine can be automatically removed
+from the provisioning network after provisioning is completed.
+
+#### Story 3
+
+As a user of Metal³, I hope that the machine can be automatically connected
+to the network I specified after provisioning is completed.
+
+#### Story 4
+
+As the administrator of Metal³, I hope that the machine can be automatically
+connected to the provisioning network before deprovisioning.
+
+#### Story 5
+
+As a user of Metal³, I hope that when the machine is removed from the cluster,
+the network configuration made to the relative network device for this machine
+will also be automatically cleared.
+
+## Design Details
+
+### Structure
+
+In the network operator, we abstract the following roles:
+
+#### Device
+
+Abstract all network devices in the cluster.
+Each type of network device must have its own CRD (eg. `Switch` CRD) and
+an interface for backend implementation.
+The CRD contains some basic information of the device
+including which `ProviderDevice` to use to interact with the device.
+The backend interface then will be implemented by `ProviderDevice`
+to use different backends to actually control the device.
+
+#### ProviderDevice
+
+`ProviderDevice` represents the method of interacting with a backend.
+It implements the backend interface of the target `Device` and contains
+all the information the backend needs to control the device.
+
+#### DevicePort
+
+Indicates a specific port. For different device types, `DevicePort` must
+have different CRDs
+(eg. `SwitchPort` CRD as `DevicePort` for `Switch` CRD as `Device`).
+
+#### Configuration
+
+Indicates the content of the specific configuration for one port.
+Like `DevicePort`, `Configuration` also must have different CRDs for
+different device types (eg. `SwitchPortConfiguration` CRD).
+
+#### ResourceLimit
+
+Indicates the limit added to a user. It is created by the administrator
+for each user to limit the use of resources (eg. the allowed VLAN range).
+Like `DevicePort`, `ResourceLimit` also must have different CRDs for
+different device types (eg. `SwitchResourceLimit` CRD).
```

Comment:
Hi @zaneb, we use the following methods to limit the vlan (or other resources) that tenants can use in a multi-tenant scenario:
1. The administrator creates ResourceLimit for each user based on the negotiation with the user
2. The controller will verify whether the VLAN to be configured is within the allowed range before configuring the switch port.
3. When users need more VLANs, they need to negotiate with the administrator, and then the administrator changes the corresponding ResourceLimit.

However, this method relies on the communication between the user and the administrator, and may be more suitable for On-Premise scenarios. In the case where the user cannot communicate with the administrator in time (such as the public cloud), it may take a long time when the user wants to use more VLANs. What do you think of this situation?

Reply:
That doesn't sound uncommon for the case where you want to increase your quota for something in most public clouds.